pandas/io/stata.py

```diff
@@ -2078,16 +2077,21 @@ def _prepare_pandas(self, data):
                 self.fmtlist[key] = self._convert_dates[key]
 
     def write_file(self):
-        self._write_header(time_stamp=self._time_stamp,
-                           data_label=self._data_label)
-        self._write_descriptors()
-        self._write_variable_labels()
-        # write 5 zeros for expansion fields
-        self._write(_pad_bytes("", 5))
-        self._prepare_data()
-        self._write_data()
-        self._write_value_labels()
-        self._file.close()
+        self._file = _open_file_binary_write(
+            self._fname, self._encoding or self._default_encoding
+        )
+        try:
+            self._write_header(time_stamp=self._time_stamp,
+                               data_label=self._data_label)
```

Comment:
Both `self._write_header` and `self._write_variable_labels()` may themselves raise a `ValueError`. The other method calls may also raise errors bubbling up from calls to other functions. In any and all cases, we want to avoid leaving `self._file` open.